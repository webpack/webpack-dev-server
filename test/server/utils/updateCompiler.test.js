'use strict';

const webpack = require('webpack');
const updateCompiler = require('../../../lib/utils/updateCompiler');
const isWebpack5 = require('../../helpers/isWebpack5');

describe('updateCompiler', () => {
  describe('simple config, no hot', () => {
    let compiler;

    beforeEach(() => {
      const webpackConfig = require('../../fixtures/simple-config/webpack.config');

      compiler = webpack(webpackConfig);
    });

    it('should apply plugins without HMR', () => {
      updateCompiler(compiler, {
        transportMode: {
          server: 'sockjs',
          client: 'sockjs',
        },
      });

      let tapsByHMR = 0;
      let tapsByProvidePlugin = 0;
      let tapsByDevServerPlugin = 0;

      compiler.hooks.compilation.taps.forEach((tap) => {
        if (tap.name === 'HotModuleReplacementPlugin') {
          tapsByHMR += 1;
        } else if (tap.name === 'ProvidePlugin') {
          tapsByProvidePlugin += 1;
        }
      });

      compiler.hooks.make.taps.forEach((tap) => {
        if (tap.name === 'DevServerPlugin') {
          tapsByDevServerPlugin += 1;
        }
      });

      expect(compiler.hooks.entryOption.taps.length).toBe(1);
      expect(tapsByHMR).toEqual(0);
      expect(tapsByProvidePlugin).toEqual(1);
      expect(tapsByDevServerPlugin).toEqual(0);
      expect(compiler.options.plugins).toHaveLength(0);
    });
  });

  describe('simple config, hot', () => {
    let compiler;

    beforeAll(() => {
      const webpackConfig = require('../../fixtures/simple-config/webpack.config');

      compiler = webpack(webpackConfig);
    });

    it('should apply plugins', () => {
      updateCompiler(compiler, {
        transportMode: {
          server: 'sockjs',
          client: 'sockjs',
        },
        hot: true,
      });

      let tapsByHMR = 0;
      let tapsByProvidePlugin = 0;
      let tapsByDevServerPlugin = 0;

      compiler.hooks.compilation.taps.forEach((tap) => {
        if (tap.name === 'HotModuleReplacementPlugin') {
          tapsByHMR += 1;
        } else if (tap.name === 'ProvidePlugin') {
          tapsByProvidePlugin += 1;
        }
      });

      compiler.hooks.make.taps.forEach((tap) => {
        if (tap.name === 'DevServerPlugin') {
          tapsByDevServerPlugin += 1;
        }
      });

      expect(compiler.hooks.entryOption.taps.length).toBe(1);
      expect(tapsByHMR).toEqual(1);
      expect(tapsByProvidePlugin).toEqual(1);
      expect(tapsByDevServerPlugin).toEqual(0);
    });
  });

  describe('simple config with HMR already, hot', () => {
    let compiler;

    beforeAll(() => {
      const webpackConfig = require('../../fixtures/simple-config/webpack.config');

      webpackConfig.plugins = [new webpack.HotModuleReplacementPlugin()];

      compiler = webpack(webpackConfig);
    });

    it('should apply plugins', () => {
      updateCompiler(compiler, {
        transportMode: {
          server: 'sockjs',
          client: 'sockjs',
        },
        hot: true,
      });

      let tapsByHMR = 0;
      let tapsByProvidePlugin = 0;
      let tapsByDevServerPlugin = 0;

      compiler.hooks.compilation.taps.forEach((tap) => {
        if (tap.name === 'HotModuleReplacementPlugin') {
          tapsByHMR += 1;
        } else if (tap.name === 'ProvidePlugin') {
          tapsByProvidePlugin += 1;
        }
      });

      compiler.hooks.make.taps.forEach((tap) => {
        if (tap.name === 'DevServerPlugin') {
          tapsByDevServerPlugin += 1;
        }
      });

      expect(compiler.hooks.entryOption.taps.length).toBe(1);
      expect(tapsByHMR).toEqual(1);
      expect(tapsByProvidePlugin).toEqual(isWebpack5 ? 4 : 1);
      expect(tapsByDevServerPlugin).toEqual(0);
      expect(compiler.options.plugins).toContainEqual(
        new webpack.HotModuleReplacementPlugin()
      );
    });
  });

  describe('multi compiler config, hot', () => {
    let multiCompiler;

    beforeAll(() => {
      const webpackConfig = require('../../fixtures/multi-compiler-2-config/webpack.config');

      webpackConfig[1].plugins = [new webpack.HotModuleReplacementPlugin()];

      multiCompiler = webpack(webpackConfig);
    });

    it('should apply plugins', () => {
      updateCompiler(multiCompiler, {
        transportMode: {
          server: 'sockjs',
          client: 'sockjs',
        },
        hot: true,
      });

      multiCompiler.compilers.forEach((compiler) => {
        let tapsByHMR = 0;
        let tapsByProvidePlugin = 0;
        let tapsByDevServerPlugin = 0;

        compiler.hooks.compilation.taps.forEach((tap) => {
          if (tap.name === 'HotModuleReplacementPlugin') {
            tapsByHMR += 1;
          } else if (tap.name === 'ProvidePlugin') {
            tapsByProvidePlugin += 1;
          }
        });

        compiler.hooks.make.taps.forEach((tap) => {
          if (tap.name === 'DevServerPlugin') {
            tapsByDevServerPlugin += 1;
          }
        });

        expect(compiler.hooks.entryOption.taps.length).toBe(1);
        expect(tapsByHMR).toEqual(1);
        expect(tapsByProvidePlugin).toEqual(1);
        expect(tapsByDevServerPlugin).toEqual(0);
      });
    });
  });
});
