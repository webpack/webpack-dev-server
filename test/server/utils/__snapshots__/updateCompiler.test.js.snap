// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateCompiler multi compiler config, hot and inline should apply plugins 1`] = `
Array [
  HotModuleReplacementPlugin {
    "fullBuildTimeout": 200,
    "multiStep": undefined,
    "options": Object {},
    "requestTimeout": 10000,
  },
]
`;

exports[`updateCompiler multi compiler config, hot and inline should apply plugins 2`] = `
Array [
  HotModuleReplacementPlugin {
    "fullBuildTimeout": 200,
    "multiStep": undefined,
    "options": Object {},
    "requestTimeout": 10000,
  },
]
`;

exports[`updateCompiler simple config with HMR already, hot and inline should apply plugins 1`] = `
Array [
  HotModuleReplacementPlugin {
    "fullBuildTimeout": 200,
    "multiStep": undefined,
    "options": Object {},
    "requestTimeout": 10000,
  },
]
`;

exports[`updateCompiler simple config, hot and inline should apply plugins 1`] = `
Array [
  HotModuleReplacementPlugin {
    "fullBuildTimeout": 200,
    "multiStep": undefined,
    "options": Object {},
    "requestTimeout": 10000,
  },
]
`;

exports[`updateCompiler simple config, inline should apply plugins without HMR 1`] = `undefined`;
